lib/index.js

```diff
@@ -28,41 +29,86 @@ const asyncResolver = ResolverFactory.createResolver({
 	extensions: [".js", ".json", ".node"],
 	fileSystem: nodeFileSystem
 });
-function resolve(context, path, request, resolveContext, callback) {
-	if (typeof context === "string") {
-		callback = resolveContext;
-		resolveContext = request;
-		request = path;
-		path = context;
-		context = nodeContext;
-	}
-	if (typeof callback !== "function") {
-		callback = resolveContext;
-	}
-	asyncResolver.resolve(context, path, request, resolveContext, callback);
-}
+
+/**
+ * @type {{
+ * (context: object, path: string, request: string, resolveContext: ResolveContext, callback: ResolveCallback): void;
+ * (context: object, path: string, request: string, callback: ResolveCallback): void;
+ * (path: string, request: string, resolveContext: ResolveContext, callback: ResolveCallback): void;
+ * (path: string, request: string, callback: ResolveCallback): void;
+ * }}
+ */
+const resolve =
+	/**
+	 * @param {object|string} context
+	 * @param {string} path
+	 * @param {string|ResolveContext|ResolveCallback} request
+	 * @param {ResolveContext|ResolveCallback=} resolveContext
+	 * @param {ResolveCallback=} callback
+	 */
+	(context, path, request, resolveContext, callback) => {
+		if (typeof context === "string") {
+			callback = /** @type {ResolveCallback} */ (resolveContext);
+			resolveContext = /** @type {ResolveContext} */ (request);
+			request = path;
+			path = context;
+			context = nodeContext;
+		}
+		if (typeof callback !== "function") {
+			callback = /** @type {ResolveCallback} */ (resolveContext);
+		}
+		asyncResolver.resolve(
+			context,
+			path,
+			/** @type {string} */ (request),
+			/** @type {ResolveContext} */ (resolveContext),
+			/** @type {ResolveCallback} */ (callback)
+		);
+	};
 
 const syncResolver = ResolverFactory.createResolver({
 	conditionNames: ["node"],
 	extensions: [".js", ".json", ".node"],
 	useSyncFileSystemCalls: true,
 	fileSystem: nodeFileSystem
 });
-function resolveSync(context, path, request) {
-	if (typeof context === "string") {
-		request = path;
-		path = context;
-		context = nodeContext;
-	}
-	return syncResolver.resolveSync(context, path, request);
-}
 
+/**
+ * @type {{
+ * (context: object, path: string, request: string): string|false;
+ * (path: string, request: string): string|false;
+ * }}
+ */
+const resolveSync =
+	/**
+	 * @param {object|string} context
+	 * @param {string} path
+	 * @param {string=} request
+	 */
+	(context, path, request) => {
+		if (typeof context === "string") {
+			request = path;
+			path = context;
+			context = nodeContext;
+		}
+		return syncResolver.resolveSync(
+			context,
+			path,
+			/** @type {string} */ (request)
+		);
+	};
+
+/** @typedef {Omit<ResolveOptions, "fileSystem"> & Partial<Pick<ResolveOptions, "fileSystem">>} ResolveOptionsOptionalFS */
```

Comment:
this thing bothers me. any suggestions welcome...

basically the `create` functions need to accept `ResolveOptions` with an optional filesystem, but everything else that takes options _requires_ a filesystem (the resolver factory etc)